docs/design-proposals/2021-05-16-scale-storage-instances-simultaneously.md

```diff
@@ -0,0 +1,72 @@
+# Scale in/out TiKV/TiFlash instances simultaneously
+
+## Summary
+
+This document presents a design to scale in/out tikv/tiflash instances simultaneously in single schedule round.
+
+## Motivation
+
+Multiple TiKV/TiFlash instances can be scaled simultaneously to speed up schedule.
+
+### Goals
+
+* Support scale multiple TiKV/TiFlash instances in one single schedule round.
+
+### Non-Goals
+
+* Scale pd or other component simultaneously.
+
+## Proposal
+
+### Scale in
+
+* Add `scaleInParallelism` in `spec.tikv` and `spec.tiflash` to specify the max instances can be scaled-in in one single schedule.
+* Default value of `scaleInParallelism` would be 1 when it's absent, for backward compatibility.
+* Extend `scaleOne` in `scaler.go` to `scaleMulti`, function signature would like:
+```javascript
+// scaleOne calculates desired replicas and delete slots from actual/desired
+// stateful sets by allowing only one pod to be deleted or created
+// it returns following values:
+// - scaling:
+//   - 0: no scaling required
+//   - 1: scaling out
+//   - -1: scaling in
+// - ordinals: pod ordinals to create or delete
+// - replicas/deleteSlots: desired replicas and deleteSlots by allowing no more than maxCount pods to be deleted or created
+func scaleMulti(actual *apps.StatefulSet, desired *apps.StatefulSet, maxCount int) (scaling int, ordinals []int32, replicas int32, deleteSlots sets.Int32)
+`````````
+* Call `scaleMulti` to get ordinals to be scaled-in, recorded as A.
+* Call PD API to get stores info, which will be used during this operation round.
+* For all ordinals waited to be scaled-in in this round:
+  * Check if number of stores with `up` state (exclude already deleted store in this round) is more than desired replicas.
```

Comment:
Not more than `more than desired replicas` but more than `Replication.MaxReplicas` in PD config.